BlueDB/src/main/java/io/bluedb/api/keys/HashGroupedKey.java

```diff
@@ -1,11 +1,67 @@
 package io.bluedb.api.keys;
 
 @SuppressWarnings("serial")
-public abstract class HashGroupedKey extends ValueKey {
+public abstract class HashGroupedKey<T extends Comparable<T>> extends ValueKey {
+
+	public abstract T getId();
+
+	@Override
+	public final int hashCode() {
+		final int prime = 31;
+		int result = 1;
+		result = prime * result + ((getId() == null) ? 0 : getId().hashCode());
+		return result;
+	}
+
+
+	@Override
+	public final String toString() {
+		return getClass().getSimpleName() + " [key=" + getId() + "]";
+	}
+
+	@Override
+	public final int compareTo(BlueKey other) {
+		if(other == null) {
+			return -1;
+		}
+		// we compare classes before grouping number because grouping number is not comparable between classes
+		int classComparison = compareClasses(other);
+		if (classComparison != 0) {
+			return classComparison;
+		}
+		int groupingNumberComparison = Long.compare(getGroupingNumber(), other.getGroupingNumber());
+		if (groupingNumberComparison != 0) {
+			return groupingNumberComparison;
+		}
+		@SuppressWarnings("unchecked")
+		HashGroupedKey<T> otherClassed = (HashGroupedKey<T>) other;
+		return getId().compareTo(otherClassed.getId());
+	}
+
 	@Override
 	public final long getGroupingNumber() {
 		long hashCodeAsLong = hashCode();
 		long integerMinAsLong = Integer.MIN_VALUE;
 		return hashCodeAsLong + Math.abs(integerMinAsLong);
 	}
+
+
+	@Override
+	public final boolean equals(Object obj) {
+		if (this == obj) {
+			return true;
+		}
+		if (obj == null) {
+			return false;
+		}
+		if (getClass() != obj.getClass()) {
+			return false;
+		}
+		@SuppressWarnings("unchecked")
+		HashGroupedKey<T> other = (HashGroupedKey<T>) obj;

```

Comment:
Just cause something is a HashGroupedKey doesn't mean that its a HashGroupedKey<T>. What if the generic types are different in the objects? Could you cast it to HashGroupedKey<?> instead?

Reply:
Look 3 lines above.  We explicitly check if the class is the same.

Reply:
Well, I guess it's a little more complicated, because you could subclass it with another generic, and instantiate two different instances.  I changed this class to generic so that subclasses (which I haven't made generic) could use final methods with standardized logic to reduce risk of bugs.  I don't have intentions of making generic subclasses.  But I think your approach is strictly superior, so I'll make the change.